dateutil/parser/_parser.py

```diff
@@ -1044,12 +1045,26 @@ def _assign_hms(self, res, value_repr, hms):
         elif hms == 2:
             (res.second, res.microsecond) = self._parsems(value_repr)
 
-    def _could_be_tzname(self, hour, tzname, tzoffset, token):
-        return (hour is not None and
-                tzname is None and
-                tzoffset is None and
-                len(token) <= 5 and
-                all(x in string.ascii_uppercase for x in token))
+    def _could_be_tzname(self, hour, tzname, tzoffset, token, tzinfos):
+        if tzinfos is None:
+            if len(token) <= 1 and token != 'Z':
+                # see GH#540
+                return False
+            return (hour is not None and
+                    tzname is None and
+                    tzoffset is None and
+                    len(token) <= 5 and
+                    all(x in string.ascii_uppercase for x in token))
+        elif isinstance(tzinfos, collections.Callable):
+            # We can't validate at this point so we let it through and validate
+            # when constructing the final return value
+            return True
+        else:
+            return token in tzinfos or (hour is not None and
+                    tzname is None and
+                    tzoffset is None and
+                    len(token) <= 5 and
+                    all(x in string.ascii_uppercase for x in token))
```

Comment:
Not sure if this is the right thing to do.  If the user _does_ pass a `tzinfos`, should we be strict in this branch and only accept `token in tzinfos` (and maybe UTC, GMT, Z or the parserinfo equivalents)?  Should the `token in tzinfo` check also require the `hour is not None and tzname is None and tzoffset is None`?